shared/src/main/java/com/couchbase/lite/Expression.java

```diff
@@ -0,0 +1,586 @@
+     */
+    public Expression isNot(Object expression) {
+        return new BinaryExpression(this, expression, BinaryExpression.OpType.IsNot);
+    }
+
+    /**
+     * Create a between expression that evaluates whether or not the current expression is
+     * between the given expressions.
+     * @param expression1 the lower bound expression.
+     * @param expression2 the upper bound expression.
+     * @return a between expression.
+     */
+    public Expression between(Object expression1, Object expression2) {
+        Expression aggr = new AggregateExpression(Arrays.asList(expression1, expression2));
+        return new BinaryExpression(this, aggr, BinaryExpression.OpType.Between);
+    }
+
+    /**
+     * Create a NOT between expression that evaluates whether or not the current expression is
+     * not between the given expressions.
+     * @param expression1 the lower bound expression.
+     * @param expression2 the upper bound expression.
+     * @return a NOT between expression.
+     */
+    public Expression notBetween(Object expression1, Object expression2) {
+        return Expression.negated(between(expression1, expression2));
+    }
+
+    /**
+     * Create an IN expression that evaluates whether or not the current expression is in the
+     * given expressions.
+     * @param expressions the expression array to evaluate with.
+     * @return an IN expression.
+     */
+    public Expression in(Object... expressions) {
+        Expression aggr = new AggregateExpression(Arrays.asList(expressions));
+        return new BinaryExpression(this, aggr, BinaryExpression.OpType.In);
+    }
+
+    /**
+     * Create a NOT IN expression that evaluates whether or not the current expression is not
+     * in the given expressions.
+     * @param expressions the expression array to evaluate with.
+     * @return a NOT IN expression.
+     */
+    public Expression notIn(Object... expressions) {
+        return Expression.negated(in(expressions));
+    }
+
+    protected abstract Object asJSON();
+
+    static class AggregateExpression extends Expression {
+        private List<Object> subexpressions;
+
+        AggregateExpression(List<Object>subexpressions) {
+            this.subexpressions = subexpressions;
+        }
+
+        @Override
+        protected Object asJSON() {
+            List<Object> json = new ArrayList<Object>();
+            for (Object exp : subexpressions) {
+                if (exp instanceof Expression)
+                    json.add(((Expression) exp).asJSON());
+                else
+                    json.add(exp);
+            }
+            return json;
+        }
+    }
+
+    static class BinaryExpression extends Expression {
+        enum OpType {
+            Add, Between, Divide, EqualTo, GreaterThan, GreaterThanOrEqualTo,
+            In, Is, IsNot, LessThan, LessThanOrEqualTo, Like, Matches,
+            Modulus, Multiply, NotEqualTo, Subtract, RegexLike,
+        }
+
+        private Object lhs;
```

Comment:
It seems lhs and rhs are commonly used. I will add comments like NSPredicate API.